runtime/src/bank.rs

```diff
@@ -3466,18 +3467,28 @@ impl Bank {
         let hash_queue = self.blockhash_queue.read().unwrap();
         let last_blockhash = hash_queue.last_hash();
         let next_durable_nonce = DurableNonce::from_blockhash(&last_blockhash);
+        let mut entry_level_dedup_map = HashSet::with_capacity(DEFAULT_NUM_TRANSACTIONS_PER_BATCH);
 
         sanitized_txs
             .iter()
             .zip(lock_results)
             .map(|(tx, lock_res)| match lock_res {
-                Ok(()) => self.check_transaction_age(
-                    tx.borrow(),
-                    max_age,
-                    &next_durable_nonce,
-                    &hash_queue,
-                    error_counters,
-                ),
+                Ok(()) => {
+                    let msg_hash = tx.borrow().message_hash();
+                    if entry_level_dedup_map.contains(msg_hash) {
+                        return Err(TransactionError::AlreadyProcessed);
+                    } else {
+                        entry_level_dedup_map.insert(msg_hash);
+                    }
```

Comment:
```suggestion
                    if !entry_level_dedup_map.insert(msg_hash) {
                        return Err(TransactionError::AlreadyProcessed);
                    }
```

`insert` returns `true` if it was a new item, false if duplicate; no need for separate contains & insert checks